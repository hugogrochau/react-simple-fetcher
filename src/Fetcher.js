import React, { PureComponent } from 'react';

export default class Fetcher extends PureComponent {
  constructor (props) {
    super(props);

    this.state = {
      fetching: true,
      data: null
    }
  }

  async componentDidMount () {
    const { handler } = this.props;

    if (typeof handler === 'function') {
      const data = await handler();

      this.setState({
        fetching: false,
        data
      });
    }
  }

  render () {
    const { component, render, children } = this.props;
    const props = this.state;

    if (component) {
      return React.createElement(component, props);
    }

    if (render) {
      return render(props);
    }

    if (typeof children === 'function') {
      return children(props);
    }

    if (children && React.Children.count(children) > 0) {
      return React.Children.only(children);
    }

    return null;
  }
}
