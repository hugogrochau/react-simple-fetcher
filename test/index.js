import React from 'react';
import renderer from 'react-test-renderer';
import Fetcher from '../src';

test('should render component prop', () => {
  const Component = props => <div {...props}>Component</div>;

  const tree = renderer.create(
    <Fetcher component={Component} />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('should render prop', () => {
  const tree = renderer.create(
    <Fetcher render={props => (
      <div {...props}>Render</div>
    )} />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('should render children when is function', () => {
  const tree = renderer.create(
    <Fetcher>
      {props => (
        <div {...props}>Children Function</div>
      )}
    </Fetcher>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('should render children', () => {
  const tree = renderer.create(
    <Fetcher>
      <div>Children</div>
    </Fetcher>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
