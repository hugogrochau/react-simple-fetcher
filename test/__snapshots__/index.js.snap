// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fetcher children> should render a child 1`] = `
<div>
  Children
</div>
`;

exports[`<Fetcher children> should render a function 1`] = `
<div>
  Children Function
</div>
`;

exports[`<Fetcher component> should render 1`] = `
<div>
  Component
</div>
`;

exports[`<Fetcher render> should render 1`] = `
<div>
  Render
</div>
`;
